parsl/dataflow/dflow.py

```diff
@@ -388,12 +380,20 @@ def launch_if_ready(self, task_id):
                     task_log_info = self._create_task_log_info(task_id, 'lazy')
                     self.monitoring.send(MessageType.TASK_INFO, task_log_info)
 
-                fu = Future()
-                fu.retries_left = 0
-                self.tasks[task_id]['exec_fu'] = fu
-                fu.set_exception(DependencyError(exceptions,
-                                                 task_id,
-                                                 None))
+                exec_fu = Future()
+                exec_fu.retries_left = 0
+                exec_fu.set_exception(DependencyError(exceptions,
+                                                      task_id,
+                                                      None))
+
+            if exec_fu:
+
+                try:
+                    exec_fu.add_done_callback(partial(self.handle_exec_update, task_id))
```

Comment:
If we had the following sequence of deps C --> B ---> A, and A were to fail, were we only triggering a dep_fail on B and not C before this fix ?

Reply:
between PR #1200 and this PR #1259, depfails didn't show up at all in any of the tests I did.

before PR #1200, depfails worked to arbitrary depth - because the results propagated through the AppFuture callback set by AppFuture.update_parent (which #1200 removed)

Reply:
before PR #1200, (and before #1257) the handle_exec_update behaviour for depfailed tasks didn't happen, though -- that's mostly irrelevant in the depfail case except for monitoring updates.

Looking at that code now, the retry logic is in that callback too, so I should test if there is a bad interaction between that and depfails.

Reply:
Related to @yadudoc 's C->B->A question, I've added a pytest case into this PR that checks 4 tasks in a chain get propagated dependency failures.

Reply:
I'm still trying to convince myself of the retries vs depfail behaviour, but I did at least generate PR #1266 on the way.

Reply:
OK, with PR #1261 merged in, which runs pytests with (amongst other excitements) retries > 0, it looks like the depfail tests introduced in 7ddd676 fail, which aligns with the suspicion I had above. I'll dig into this more.

Reply:
I am comfortable with the dep_fail handling introduced in 4a84064 which passes the tests now